References infrastructure

BobBorges opened the pull request:
A bibtex solution for managing references within the corpus.

MansMeg commented on .gitignore:
Why do we add this?

BobBorges replied:
It's not related to references, but you'll find the same in other PRs that haven't been merged to dev yet.

It's to more easily ignore things that are under development and not ready to commit, or things that we never want to commit but don't necessarily want to delete from the local disk. After a couple months working with this, running `git status` fills the whole screen height with untracked files. With this addition, I keep e.g. `_scripts` locally with files that I don't want committed (yet), and just move them over to `scripts` when they are to be committed.

MansMeg replied:
I see but then you want to have this locally, not in the corpus, right? You have a whole list because you work with the corpus, no?

BobBorges replied:
> You have a whole list

My addition here is only line 14. I added line 13, but in an older PR that was already merged.

> I see but then you want to have this locally, not in the corpus, right? 

In my view it's on a par with other things that we want to ignore in the whole repository. It can be done locally, but in a slightly more cumbersome way. Do you see this causing problems down the line?

MansMeg replied:
Nah. I just dint understand why. But ok, lets do this. No big thing.

MansMeg commented on corpus/references/README.md:
```diff
@@ -0,0 +1,65 @@
+# References
+
+This directory contains references used within the corpus. The sub-directory `bibfiles/` contains `.bib` files for each reference. The file `references.bib` is a compiled single reference file for the `.bib` files contained in the `bibfiles/` directory. `references.bib` is compiled using `scripts/references/compile-bib-references.py`.
```
We should not have the same data in multiple files. Is it sufficient to just have the references bib?

BobBorges replied:
I agree that we shouldn't have duplicate data. A `references.bib` file would be sufficient for the corpus. *But* I don't think it's a good idea to add directly to that file (if you've ever had the pleasure of searching for error in 1000s of lines of bibtex, you know) -- having individual bib files is an easy way to manage additions and verify valid bibtex when compiling the `references.bib` file. Would you be happy to move the `bibfiles/` dir out of `corpus/'?

MansMeg replied:
Ok. Then maybe we keep the references as separate files and ad your combination script to the library? Then we can combine it if needed and only store separate bibtex files? I just want to avoid duplicated data being commited.

MansMeg commented on scripts/references/compile-bib-references.py:
We should avoid having code in the corpus. The corpus should be the state and there should be code in another folder (or other repo) that is used.

BobBorges replied:
This is in `scripts/`.

MansMeg replied:
Ah great! But if we are supposed to use it more frequently maybe it should be in the library?

BobBorges replied:
I can move the functions to the module, but you usually don't call py libraries directly, so there has to be a script somewhere anyway. If it were only up to me, I would only be adding code to the module that gets used multiple times, so in this script for example, I would only add `validate_bib()` to the module if it will be needed in another script -- if it's only in the compile script, then leave it as a function in the script.

MansMeg replied:
Ok. Fair enough.

MansMeg commented:
I have some comments (see above). I think it is easier to just have one bib-file for now or all separate bibfiles. Then we could add functionality to the python library to combine them. But we should not have the same data stored multiple times in the data and code should be in the python library as a function that users can use.

BobBorges commented:
Last commit should satisfy criticism:
• removed the compiled .bib file
• added an arg to the compile script to define the output file & set default to an untracked file (+1 for **/_*)

MansMeg commented:
Yes! Then I'm happy. Maybe update the README to reflect this change (if not done already). Then I'm happy. So if @BobBorges gives a green light on this I think @ninpnin can merge.

MansMeg commented:
Boom - you have a file to resolve as well.

BobBorges commented:
of course the actual documentation :|

MansMeg commented:
Great! I'm happy now.

ninpnin commented:
LGTM
